### PR TITLE
gh-127271: Replace use of PyCell_GET/SET

### PR DESCRIPTION
**Pull request by nascheme**
These macros are not safe to use in the free-threaded build.  Use `PyCell_GetRef()` and `PyCell_SetTakeRef()` instead. Add `PyCell_GET` to the free-threading howto table of APIs that return borrowed refs. Add critical sections to `PyCell_GET` and `PyCell_SET`.

Fix some thread safety issues with cell objects: `cell_richcompare`, `cell_repr`, `cell_set_contents` needed some work.  Add an addtional unit test module that tries to trigger data races for these.

* Issue: gh-127271

### CODE REVIEW

**Comment by nascheme** [comment]
While tinkering with the cellobject code, I found a couple more data races.  I've included the fixes for those in this PR.  

For the reviewer, the reference counting changes to `super_init_without_args` are quite tricky and I'm not so confident that I didn't introduce a bug in one of the many infrequently taken branches.  Most of the other changes are pretty simple.  It's mostly about using strong references rather than borrowed ones.  I don't think there should be any significant performance impact in the non-free-threaded build (couple extra incref/decref pairs but mostly in error handling branches).

File: Lib/test/test_free_threading/test_races.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    pass
+
+
+def do_race(func1, func2):
+    """Run func1() and func2() repeatedly in separate threads."""
+    n = 1000
+
+    def repeat(func):
+        for _i in range(n):
+            func()
```
Comment (mpage) [inline Lib/test/test_free_threading/test_races.py]: I'd recommend using a `threading.Barrier` at the beginning of `repeat` to ensure that each thread has arrived at the same point before we attempt to exercise races. As written, it's possible for one thread to finish execution before the other starts (due to some combination `n` being small enough, `func` running quickly, and the whims of the scheduler).

Something like:

```py
def repeat(func, barrier):
        barrier.wait()
        for _i in range(n):
            func()
```

**Review by mpage — COMMENTED** [review]
I kicked off benchmark runs for this change for both the default and free-threaded builds. I'll post the results here once they're ready, likely tomorrow morning.

**Comment by mpage** [comment]
Performance on the default build looks [neutral](https://github.com/facebookexperimental/free-threading-benchmarking/tree/main/results/bm-20241127-3.14.0a2+-b09c4cf).

**Comment by nascheme** [comment]
I can't reproduce those two failing tests on my Debian dev machine.  My guess is the failure is not related to this commit.  I'll investigate some more though.

**Pull request merged**